src/main/java/org/apache/maven/plugins/shade/resource/SisuIndexResourceTransformer.java

```diff
@@ -0,0 +1,130 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Scanner;
+import java.util.jar.JarEntry;
+import java.util.jar.JarOutputStream;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.maven.plugins.shade.relocation.Relocator;
+
+/**
+ * Resource transformer that relocates classes in {@code META-INF/sisu/javax.inject.Named} and appends resources
+ * into a single resource.
+ *
+ * @since 3.3.0
+ */
+public class SisuIndexResourceTransformer
+    extends AbstractCompatibilityTransformer
+{
+    private static final String SISU_INDEX_PATH = "META-INF/sisu/javax.inject.Named";
+
+    private static final String NEWLINE = "\n";
+
+    private ServiceStream serviceStream;
+
+    private long time = Long.MIN_VALUE;
+
+    @Override
+    public boolean canTransformResource( final String resource )
+    {
+        return resource.equals( SISU_INDEX_PATH );
+    }
+
+    @Override
+    public void processResource( final String resource,
+                                 final InputStream is,
+                                 final List<Relocator> relocators,
+                                 long time ) throws IOException
+    {
+        if ( serviceStream == null )
+        {
+            serviceStream = new ServiceStream();
+        }
+
+        final String content = IOUtils.toString( is, StandardCharsets.UTF_8 );
+        Scanner scanner = new Scanner( content );
+        while ( scanner.hasNextLine() )
+        {
+            String relContent = scanner.nextLine();
+            for ( Relocator relocator : relocators )
+            {
+                if ( relocator.canRelocateClass( relContent ) )
+                {
+                    relContent = relocator.applyToSourceContent( relContent );
+                }
+            }
+            serviceStream.append( relContent );
+            serviceStream.append( NEWLINE );
+        }
+
+        if ( time > this.time )
+        {
+            this.time = time;        
+        }
+    }
+
+    @Override
+    public boolean hasTransformedResource()
+    {
+        return serviceStream != null;
+    }
+
+    @Override
+    public void modifyOutputStream( final JarOutputStream jos )
+        throws IOException
+    {
+        JarEntry jarEntry = new JarEntry( SISU_INDEX_PATH );
+        jarEntry.setTime( time );
+        jos.putNextEntry( jarEntry );
+        IOUtils.copy( serviceStream.toInputStream(), jos );
+        jos.flush();
+        serviceStream.reset();
+   }
+
+    static class ServiceStream
```

Comment:
I suspect there's a simpler way to accomplish this, likely with code in the JDK or perhaps commons.io. Yes, I think you can forget about copying everything into streams. Just build a string when you transform and write that into the jar.

Reply:
agreed